### PR TITLE
persist: begin hooking up PersistConfig to LaunchDarkly

### LINKED ISSUE #16753: Add dynamic configuration for PersistConfig
As outlined in the [user stories section of the design doc for dynamic environment configuration](https://www.notion.so/materialize/Dynamic-environment-configuration-ad11421b40934e4cba675bbe550cc2b2#80ee54c75ba448eaa23c1b5b81150f35), the storage team would like to be able to tweak [PersistConfig options](https://github.com/MaterializeInc/materialize/blob/0807a06845cce08dd1b9e88b8f4fe7b743bf71ed/src/persist-client/src/lib.rs#L186-L242) using dynamic configuration options.

## Tasks

- [x] compute: Add LaunchDarkly parameter for persist config
    #17076
- [x] compute: Plumb persist config param through to controllers
    #17076
- [ ] storage: Use dynamic options to configure persist clients

## Description

As discussed offline with @teskje and @danhhz, the compute team will put all the required plumbing in place for storage to use the configuration options.

Dan confirms `PersistClient` should be able to handle differing configs (let's say one used in a `persist_source` and another one in a `persist_sink`) which means we don't have to introduce any additional consistency constraints for distribution of these options.

### PR DESCRIPTION
**Pull request by danhhz**
This introduces DynamicConfig, which is a new abstraction around
configuration that can change on a lifetime shorter than a process. It
also hooks up blob_target_size end-to-end between LaunchDarkly and
PersistConfig (huge ty to Jan for doing all the hard bits here).

All configs where the usage currently adjusted dynamically are moved to
DynamicConfig. However, intentionally left out of scope:
- Hooking up compaction_minimum_timeout up end-to-end. Handling dynamic
  configs of type Duration has some tradeoffs we still need to explore.
- Making Vars and LaunchDarkly entries for every persist config; this is
  entirely boilerplate but it's a _lot_ of boilerplate

This hooks up the last mile of LaunchDarkly->PersistConfig and thus...

Closes #16753

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/persist-client/src/internal/compact.rs
Comment (pH14) [inline src/persist-client/src/internal/compact.rs:200]: how were we thinking of setting these in LD? are we able to atomically update multiple values, or is it possible we'd read these 3 knobs with arbitrary skew

File: src/persist-client/src/lib.rs
Comment (pH14) [inline src/persist-client/src/lib.rs:377]: supernit: might be nice to separate out the idea of dynamic config from the LD implementation details

**Review by pH14 — COMMENTED** [review]
Makes sense to me! This largely matches the picture I had in my head for this problem, all seems quite reasonable.

File: src/persist-client/src/internal/compact.rs
Comment (danhhz) [reply src/persist-client/src/internal/compact.rs:200]: great q, dunno! if we can handle the latter, certainly seems less brittle to bugs in the config propagation (and we could probably relax these SeqCsts once we thought about Ordering a bit, whereas atomic updates would require going _more_ expensive and having something like a mutex around all the dynamic bits). i'd love it if we could make these functions something that are safe to call in hot-ish loops (fewer footguns that way)

Comment (pH14) [reply src/persist-client/src/internal/compact.rs:200]: yeah I like the idea of relaxing the orderings. if we can't get tight guarantees from LD anyway, we might as well minimize the cost of reading the values in mem

**Comment by benesch** [comment]
What don't you like about how this turned out, out of curiosity? Seems pretty nice to me!

**Comment by danhhz** [comment]
> What don't you like about how this turned out, out of curiosity? Seems pretty nice to me!

oh, nothing fundamental! snapshotting the `blob_target_size` to pass down in compaction and the extra`cfg.dynamic` level of indirection both seem clunky to me. both are probably just a matter of some code golf that I didn't have time for when getting this draft out

**Comment by benesch** [comment]
Oh stuff on the `persist` side, rather than the LaunchDarkly side! Sweet! 👍🏽

File: src/persist-client/src/internal/compact.rs
Comment (teskje) [reply src/persist-client/src/internal/compact.rs:200]: I don't think it would be possible to atomically update multiple parameters in LD, at least not if those are represented as separate flags. What we could do is represent them as a single text/json flag in LD, do the decoding in environmentd, and ensure that our code always propagates them as one.

That would be less ergonomic from an LD-user perspective and more of a hassle to implement, so if persist could be taught to handle skewed individual updates instead, that would definitely be nicer!

Comment (danhhz) [reply src/persist-client/src/internal/compact.rs:200]: Sounds like we have our answer!

**Comment by pH14** [comment]
> snapshotting the blob_target_size to pass down in compaction and the extracfg.dynamic level of indirection both seem clunky to me. both are probably just a matter of some code golf that I didn't have time for when getting this draft out

Heh I actually liked both of these! It's important to recognize when a value is and is not dynamic in case some logic expects/requires the value to be unchanging after it's read, so IMO it's nice to have a separate `.dynamic` namespace that makes this more obvious. If anything, I think we could eventually lean even harder on the types to make it (close to) impossible to mistakenly reuse a dynamic value when it's unsafe to do so.

File: src/persist-client/src/internal/compact.rs
Comment (danhhz) [reply src/persist-client/src/internal/compact.rs:200]: Documented this as part of the new apply method.

File: src/persist-client/src/lib.rs
Comment (danhhz) [reply src/persist-client/src/lib.rs:377]: done!

**Review by danhhz — COMMENTED** [review]
Ended up rewriting this, so best to review from scratch. (Though, the design is very similar.) First two commits are pure refactors I pulled out into #17330

**Comment by danhhz** [comment]
Rebased on top of the refactor PR so this should be ready to go!

File: src/persist-client/src/cfg.rs
Comment (teskje) [inline src/persist-client/src/cfg.rs:413]: If it's hard to apply `compaction_minimum_timeout` dynamically, how about removing it as a `SystemVar` and `PersistParameter` for now? It served its purpose as a test case for having duration parameters, I don't think we need it for that anymore. On the other hand, it is confusing to have a flag in Launch Darkly that doesn't have an effect when changed.

**Review by teskje — APPROVED** [review]
Looks good to me, FWIW.

File: src/persist-client/src/cfg.rs
Comment (teskje) [reply src/persist-client/src/cfg.rs:413]: Ah, I wasn't clear on how big of a lift removing this TODO was. I'm fine with leaving the `SystemVar` unimplemented for a bit longer!

**Comment by danhhz** [comment]
just realized this doesn't include the very critical step of actually calling the new `apply` method, whoops. hooking that up nicely requires a minor refactor, so I'll do that first and then rebase this again once it merges

**Comment by danhhz** [comment]
> just realized this doesn't include the very critical step of actually calling the new apply method, whoops. hooking that up nicely requires a minor refactor, so I'll do that first and then rebase this again once it merges

and done!

**Review by pH14 — COMMENTED** [review]
I think I'm a little unclear on what should and should not be hooked up at this point. From my read on this, `PersistParameters::apply` will update a particular copy of `PersistConfig` in the client cache, but we clone `PersistConfig` left and right throughout persist so it doesn't seem like the change would ripple to those. Will we need to change how we pass around/reference `PersistConfig`?

The code movement overall seems fine to me, though I wouldn't mind a comment on the difference between `PersistConfig` and `PersistParameters` somewhere, since their names alone don't explain the different intents

**Comment by danhhz** [comment]
> From my read on this, PersistParameters::apply will update a particular copy of PersistConfig in the client cache, but we clone PersistConfig left and right throughout persist so it doesn't seem like the change would ripple to those. Will we need to change how we pass around/reference PersistConfig?

All PersistConfigs that come out of the same PersistClientCache (which is all of them, we have one PersistClientCache per process) share the same `Arc<DynamicConfig>`

**Comment by danhhz** [comment]
> The code movement overall seems fine to me, though I wouldn't mind a comment on the difference between PersistConfig and PersistParameters somewhere, since their names alone don't explain the different intents

Oh turns out I already did write this comment on the rustdoc for PersistParameters, though obviously it didn't do its job. Can I ask where you'd have expected it to be? I'll move the comment there (and expand on it).

```
/// These reflect updates made to LaunchDarkly. They're passed from environmentd
/// through storage and compute commands and applied to PersistConfig to change
/// its values.
```

**Comment by pH14** [comment]
> Oh turns out I already did write this comment on the rustdoc for PersistParameters, though obviously it didn't do its job. Can I ask where you'd have expected it to be? I'll move the comment there (and expand on it).

Yeah GitHub was folding this section away so I didn't see it 🙈

> All PersistConfigs that come out of the same PersistClientCache (which is all of them, we have one PersistClientCache per process) share the same Arc<DynamicConfig>

Oh! Excellent, that makes way more sense then. LGTM modulo any further changes to docs

**Review by pH14 — APPROVED** [review]

**Comment by danhhz** [comment]
Nightlies: https://buildkite.com/materialize/nightlies/builds/1776

> Sounds like there's some other opportunities for more commentary here, too

Done!

**Comment by danhhz** [comment]
After retrying a bunch of jobs on the nighties run that got scheduled on a bad agent, we were left with 2 failures, both of which were happening in main at the point this was rebased to. I double checked and both were failing with the same `reference to absent collection: CollectionMissing(User(358))` issue that was happening on main. It's since been fixed, so if this was a scarier change, I'd probably pull and rerun nightlies to be sure, but the only real expected difference here is an AtomicUsize load where we previously had just a usize, so that seems like overkill.

TFTRs!

**Pull request merged**
